src/index.ts

```diff
@@ -0,0 +1,123 @@
+import * as _ from 'lodash'
+import * as tar from 'tar-stream'
+
+import Bundle from './bundle'
+import { FileInfo } from './fileInfo'
+import { Resolver } from './resolver'
+import * as Utils from './utils'
+
+// Import some default resolvers
+import DockerfileResolver from './resolvers/dockerfile'
+import DockerfileTemplateResolver from './resolvers/dockerfileTemplate'
+import ArchDockerfileResolver from './resolvers/archDockerfile'
+
+// re-export
+export { ArchDockerfileResolver,
+         Bundle,
+				 DockerfileResolver,
+         DockerfileTemplateResolver,
+         FileInfo,
+         Resolver }
+
+export interface ResolvedBundle {
+	projectType: string,
+	tarStream: tar.Pack
+}
+
+export function resolveBundle(bundle: Bundle, resolvers: Resolver[]): Promise<ResolvedBundle> {
+	return new Promise<ResolvedBundle>((resolve: (resolved: ResolvedBundle) => void,
+                                		  reject: (err: Error) => void) => {
+		const extract = tar.extract()
+		const pack = tar.pack()
+
+		extract.on('entry', (header: tar.TarHeader, stream: NodeJS.ReadableStream, next: () => void) => {
+			// Read the contents into a buffer
+			let buf: string = ''
+
+			stream.on('data', (data: string) => {
+				buf += data
+			})
+
+			stream.on('end', () => {
+				// Send the file along to the next tar stream regardless
+				pack.entry(header, buf)
+
+				// create a FileInfo from the header
+				const info: FileInfo = {
+					name: Utils.normalizeTarEntry(header.name),
+					size: header.size,
+					contents: buf
+				}
+
+				// Now provide the resolvers with the information and file
+				resolvers.map((resolver) => {
+					resolver.entry(info)
+				})
+			})
+
+			next()
+		})
+
+		extract.on('finish', () => {
+			let maybeResolver: Resolver | undefined = undefined
+			// Check to see if any of the resolvers are happy, in `priority` order
+			_(_.orderBy(resolvers, (val: Resolver) => val.priority, ['desc']))
+			.each((resolver: Resolver) => {
+				resolver.provideBundle(bundle)
+				if (resolver.isSatisfied()) {
+					maybeResolver = resolver
+					// Stop iterating early
+					return false
+				}
+				return true
+			})
+
+			// if no resolver was happy this is an error
+			if (maybeResolver === undefined) {
+				reject(new Error('No project type resolution could be performed'))
+				resolvers.map((resolver) => resolver.reset())
+				return
+			}
+
+			const resolver: Resolver = maybeResolver
+
+			// Now that we have a resolver, add the new files needed to the stream
+			resolver.resolve()
+			.then((additionalItems) => {
+				additionalItems.map((file: FileInfo) => {
+					pack.entry({ name: file.name, size: file.size }, file.contents)
+				})
+
+				// all of the extra files have now been added to the stream, resolve
+				// the promise with it
+				pack.finalize()
+
+				resolve({
+					projectType: resolver.name,
+					tarStream: pack
+				})
+
+				// reset the resolvers for next time
+				resolvers.map((resolver) => resolver.reset())
```

Comment:
Even if we do want to require `reset` (see comment below) this feels buggy, because we resolve the promise before we've reset the resolvers. I think we get away with it because the next promise step runs async (though I'm not sure if that's _always_ true), but it'd be better to resolve after resetting, and keep this clearer and safer.